File: tensorcast/common/capability_token.py
#  Copyright (c) 2026, TensorCast Team.

from __future__ import annotations

import hashlib
import hmac
from dataclasses import dataclass

from google.protobuf.message import Message

from tensorcast.proto.common.v1 import capability_token_pb2


@dataclass(frozen=True, slots=True)
class CapabilityTokenKey:
    version: int
    secret: bytes


@dataclass(frozen=True, slots=True)
class CapabilityTokenConfig:
    active: CapabilityTokenKey
    previous: tuple[CapabilityTokenKey, ...] = ()


def _serialize_deterministic(message: Message) -> bytes:
    return message.SerializeToString(deterministic=True)


def serialize_scope(message: Message) -> bytes:
    return _serialize_deterministic(message)


class CapabilityTokenManager:
    def __init__(self, config: CapabilityTokenConfig) -> None:
        self._config = config

    def configured(self) -> bool:
        return bool(self._config.active.version and self._config.active.secret)

    def mint(
        self,
        *,
        issuer_daemon_id: str,
        audience: capability_token_pb2.CapabilityAudience,
        scope: bytes,
        expires_at_ms: int,
        queue_epoch: capability_token_pb2.QueueEpochFencing | None = None,
    ) -> bytes:
        if not self.configured():
            raise ValueError("capability token keys are not configured")
        if not issuer_daemon_id:
            raise ValueError("issuer_daemon_id is required")
        if audience == capability_token_pb2.CAPABILITY_AUDIENCE_UNSPECIFIED:
            raise ValueError("audience is required")
        if not scope:
            raise ValueError("scope is required")
        if expires_at_ms <= 0:
            raise ValueError("expires_at_ms is required")

        env = capability_token_pb2.CapabilityTokenEnvelope(
            token_version=int(self._config.active.version),
            issuer_daemon_id=str(issuer_daemon_id),
            audience=audience,
            scope=bytes(scope),
            expires_at_ms=int(expires_at_ms),
        )
        if queue_epoch is not None:
            env.queue_epoch.CopyFrom(queue_epoch)
        auth = self._compute_auth_tag(env, self._config.active.secret)
        env.auth_tag = auth
        return _serialize_deterministic(env)

    def verify(
        self,
        token: bytes,
        *,
        expected_audience: capability_token_pb2.CapabilityAudience,
        expected_issuer: str,
        now_ms: int,
        require_not_expired: bool,
    ) -> capability_token_pb2.CapabilityTokenEnvelope:
        if not self.configured():
            raise ValueError("capability token keys are not configured")
        env = capability_token_pb2.CapabilityTokenEnvelope()
        env.ParseFromString(token)
        if (
            env.token_version == 0
            or not env.issuer_daemon_id
            or env.audience == capability_token_pb2.CAPABILITY_AUDIENCE_UNSPECIFIED
            or not env.scope
            or not env.auth_tag
        ):
            raise ValueError("capability token missing required fields")
        if expected_issuer and env.issuer_daemon_id != expected_issuer:
            raise PermissionError("capability token issuer mismatch")
        if (
            expected_audience != capability_token_pb2.CAPABILITY_AUDIENCE_UNSPECIFIED
            and env.audience != expected_audience
        ):
            raise PermissionError("capability token audience mismatch")
        if require_not_expired and env.expires_at_ms <= int(now_ms):
            raise PermissionError("capability token expired")

        key = self._key_for_version(env.token_version)
        if key is None:
            raise PermissionError("capability token key version not recognized")
        expected = self._compute_auth_tag(env, key.secret)
        if not hmac.compare_digest(expected, env.auth_tag):
            raise PermissionError("capability token auth tag mismatch")
        return env

    @staticmethod
    def looks_like_envelope(token: bytes) -> bool:
        env = capability_token_pb2.CapabilityTokenEnvelope()
        try:
            env.ParseFromString(token)
        except Exception:  # noqa: BLE001
            return False
        return bool(
            env.token_version
            and env.issuer_daemon_id
            and env.audience != capability_token_pb2.CAPABILITY_AUDIENCE_UNSPECIFIED
            and env.scope
            and env.auth_tag
        )

    def _key_for_version(self, version: int) -> CapabilityTokenKey | None:
        if version == self._config.active.version:
            return self._config.active
        for key in self._config.previous:
            if version == key.version:
                return key
        return None

    @staticmethod
    def _compute_auth_tag(
        env: capability_token_pb2.CapabilityTokenEnvelope, secret: bytes
    ) -> bytes:
        unsigned_env = capability_token_pb2.CapabilityTokenEnvelope()
        unsigned_env.CopyFrom(env)
        unsigned_env.auth_tag = b""
        payload = _serialize_deterministic(unsigned_env)
        return hmac.new(secret, payload, hashlib.sha256).digest()

File: core/common/capability_token.cc
// Copyright (c) 2026, TensorCast Team.

#include "core/common/capability_token.h"

#include <cstring>

#include "absl/status/status.h"
#include "absl/strings/string_view.h"
#include "google/protobuf/io/coded_stream.h"
#include "google/protobuf/io/zero_copy_stream_impl_lite.h"
#include "openssl/hmac.h"
#include "openssl/sha.h"

namespace tensorcast::common {
namespace {

constexpr size_t kAuthTagBytes = SHA256_DIGEST_LENGTH;

absl::StatusOr<std::string> serialize_deterministic(const google::protobuf::Message& message) {
  const size_t size = message.ByteSizeLong();
  std::string out;
  out.resize(size);
  google::protobuf::io::ArrayOutputStream aos(out.data(), static_cast<int>(out.size()));
  google::protobuf::io::CodedOutputStream cos(&aos);
  cos.SetSerializationDeterministic(true);
  if (!message.SerializeToCodedStream(&cos)) {
    return absl::InternalError("failed to serialize protobuf deterministically");
  }
  return out;
}

} // namespace

CapabilityTokenManager::CapabilityTokenManager(CapabilityTokenConfig config) : config_(std::move(config)) {}

bool CapabilityTokenManager::configured() const {
  return config_.active.version != 0 && !config_.active.secret.empty();
}

absl::StatusOr<std::string> CapabilityTokenManager::serialize_scope_deterministic(
    const google::protobuf::Message& message) {
  return serialize_deterministic(message);
}

absl::StatusOr<std::string> CapabilityTokenManager::compute_auth_tag(
    const tensorcast::common::v1::CapabilityTokenEnvelope& envelope,
    std::string_view secret) const {
  if (secret.empty()) {
    return absl::InvalidArgumentError("capability token secret is empty");
  }
  tensorcast::common::v1::CapabilityTokenEnvelope unsigned_env = envelope;
  unsigned_env.clear_auth_tag();
  auto payload_or = serialize_deterministic(unsigned_env);
  if (!payload_or.ok()) {
    return payload_or.status();
  }
  unsigned int out_len = 0;
  unsigned char out[kAuthTagBytes];
  const unsigned char* result = HMAC(
      EVP_sha256(),
      secret.data(),
      static_cast<int>(secret.size()),
      reinterpret_cast<const unsigned char*>(payload_or->data()),
      payload_or->size(),
      out,
      &out_len);
  if (result == nullptr || out_len != kAuthTagBytes) {
    return absl::InternalError("failed to compute capability token auth tag");
  }
  return std::string(reinterpret_cast<const char*>(out), out_len);
}

absl::StatusOr<std::string> CapabilityTokenManager::mint(
    std::string_view issuer_daemon_id,
    tensorcast::common::v1::CapabilityAudience audience,
    std::string_view scope_bytes,
    uint64_t expires_at_ms,
    const tensorcast::common::v1::QueueEpochFencing* queue_epoch) const {
  if (!configured()) {
    return absl::FailedPreconditionError("capability token keys are not configured");
  }
  if (issuer_daemon_id.empty()) {
    return absl::InvalidArgumentError("issuer_daemon_id is required");
  }
  if (audience == tensorcast::common::v1::CAPABILITY_AUDIENCE_UNSPECIFIED) {
    return absl::InvalidArgumentError("capability token audience is required");
  }
  if (scope_bytes.empty()) {
    return absl::InvalidArgumentError("capability token scope is required");
  }
  if (expires_at_ms == 0) {
    return absl::InvalidArgumentError("capability token expires_at_ms is required");
  }

  tensorcast::common::v1::CapabilityTokenEnvelope envelope;
  envelope.set_token_version(config_.active.version);
  envelope.set_issuer_daemon_id(std::string(issuer_daemon_id));
  envelope.set_audience(audience);
  envelope.set_scope(std::string(scope_bytes));
  envelope.set_expires_at_ms(expires_at_ms);
  if (queue_epoch != nullptr) {
    *envelope.mutable_queue_epoch() = *queue_epoch;
  }

  auto auth_or = compute_auth_tag(envelope, config_.active.secret);
  if (!auth_or.ok()) {
    return auth_or.status();
  }
  envelope.set_auth_tag(*auth_or);

  return serialize_deterministic(envelope);
}

absl::StatusOr<tensorcast::common::v1::CapabilityTokenEnvelope> CapabilityTokenManager::verify(
    std::string_view token_bytes,
    tensorcast::common::v1::CapabilityAudience expected_audience,
    std::string_view expected_issuer,
    absl::Time now,
    bool require_not_expired) const {
  if (!configured()) {
    return absl::FailedPreconditionError("capability token keys are not configured");
  }
  tensorcast::common::v1::CapabilityTokenEnvelope envelope;
  if (!envelope.ParseFromArray(token_bytes.data(), static_cast<int>(token_bytes.size()))) {
    return absl::InvalidArgumentError("capability token parse failed");
  }
  if (envelope.token_version() == 0 || envelope.issuer_daemon_id().empty() ||
      envelope.audience() == tensorcast::common::v1::CAPABILITY_AUDIENCE_UNSPECIFIED || envelope.scope().empty() ||
      envelope.auth_tag().empty()) {
    return absl::InvalidArgumentError("capability token missing required fields");
  }
  if (!expected_issuer.empty() && envelope.issuer_daemon_id() != expected_issuer) {
    return absl::PermissionDeniedError("capability token issuer mismatch");
  }
  if (expected_audience != tensorcast::common::v1::CAPABILITY_AUDIENCE_UNSPECIFIED &&
      envelope.audience() != expected_audience) {
    return absl::PermissionDeniedError("capability token audience mismatch");
  }

  const uint64_t now_ms = static_cast<uint64_t>(absl::ToUnixMillis(now));
  if (require_not_expired && envelope.expires_at_ms() <= now_ms) {
    return absl::PermissionDeniedError("capability token expired");
  }

  std::string secret;
  if (envelope.token_version() == config_.active.version) {
    secret = config_.active.secret;
  } else {
    for (const auto& key : config_.previous) {
      if (key.version == envelope.token_version()) {
        secret = key.secret;
        break;
      }
    }
  }
  if (secret.empty()) {
    return absl::PermissionDeniedError("capability token key version not recognized");
  }

  auto auth_or = compute_auth_tag(envelope, secret);
  if (!auth_or.ok()) {
    return auth_or.status();
  }
  if (auth_or->size() != envelope.auth_tag().size() ||
      std::memcmp(auth_or->data(), envelope.auth_tag().data(), auth_or->size()) != 0) {
    return absl::PermissionDeniedError("capability token auth tag mismatch");
  }
  return envelope;
}

bool CapabilityTokenManager::looks_like_envelope(std::string_view token_bytes) {
  tensorcast::common::v1::CapabilityTokenEnvelope envelope;
  if (!envelope.ParseFromArray(token_bytes.data(), static_cast<int>(token_bytes.size()))) {
    return false;
  }
  if (envelope.token_version() == 0 || envelope.issuer_daemon_id().empty() ||
      envelope.audience() == tensorcast::common::v1::CAPABILITY_AUDIENCE_UNSPECIFIED || envelope.scope().empty() ||
      envelope.auth_tag().empty()) {
    return false;
  }
  return true;
}

} // namespace tensorcast::common
